Recruit-Training/Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp

```diff
@@ -16,22 +25,75 @@ BatteryStateOfChargeService::~BatteryStateOfChargeService()
 
 double BatteryStateOfChargeService::totalAmpHoursUsed() const
 {
-    return 0.0;
+    //double initialAmpHours = initialStateOfChargePercent_ * BATTERY_AMP_HOUR_CAPACITY;
+
+    return BATTERY_AMP_HOUR_CAPACITY - remainingAmpHours;
 }
 
 bool BatteryStateOfChargeService::isCharging() const
 {
-    return false;
+    if(data_.current < 0) return true;
+    return false;   //assuming positive current is not charging
 }
 
 QTime BatteryStateOfChargeService::timeWhenChargedOrDepleted() const
 {
-    return QTime::currentTime();
+    double hoursTill;
+    if(isCharging()){
+        hoursTill = 0.0 - (BATTERY_AMP_HOUR_CAPACITY - remainingAmpHours) / data_.current;
+    }
+    else{
+        hoursTill = remainingAmpHours / data_.current;
+    }
+
+    int milliseconds = hoursTill * 3600000; int remainingMilis;
+    int hours = milliseconds / 3600000 % 24; remainingMilis = milliseconds % 3600000;
+    int minutes = remainingMilis / 60000 % 60; remainingMilis = milliseconds % 60000;
+    int seconds = remainingMilis / 1000 % 60; remainingMilis = milliseconds % 1000;
+    int mili = remainingMilis;
+
+    return QTime(hours, minutes, seconds, mili);
```

Comment:
There's a function that takes in the number of msecs and adds it to a QTime. See if you can use that instead to reduce the number of computations needed.